Cargo.toml

```diff
@@ -28,3 +28,8 @@ resolver = "2"
 [profile.bench]
 codegen-units = 1
 lto = true
+
+[patch.crates-io]
+# TODO(XXX): Remove this once 0.25 is released - we want the ECH fixes from
+#            https://github.com/hickory-dns/hickory-dns/pull/2183
+hickory-resolver = { git = "https://github.com/hickory-dns/hickory-dns", rev = "6334a01430088ead8642cafaee592ec7bf49831f" }
```

Comment:
I _think_ the MSRV job is upset with this because this states a requirement of 1.67? This was fine without a [patch], because `connect-tests` is not part of the MSRV check job usually. But for whatever reason, cargo decides it needs to re-find this dependency to build something unrelated.

Reply:
I'm confused about what's going on with this CI task. Your theory about the `hickory-dns` MSRV seems plausible, but the output looks more like it can't resolve the patch content, not that the MSRV was incompatible.

I can reproduce locally w/ the 1.63 toolchain but haven't been able to figure out a fix yet.

Reply:
It does seem to work locally w/ 1.64, 1.65, 1.66, and 1.67.

Maybe there's a bug with patch resolution in 1.63?

Reply:
I wonder if it's related to 1.64+ having added [workspace inheritance](https://blog.rust-lang.org/2022/09/22/Rust-1.64.0.html#cargo-improvements-workspace-inheritance-and-multi-target-builds) since the hickory-dns project is using that: https://github.com/hickory-dns/hickory-dns/blob/f1489da675c21fddc189f2c9505bc9da6c156835/Cargo.toml#L29-L35

Reply:
> I wonder if it's related to 1.64+ having added workspace inheritance

I think that's the case unfortunately and I can't think of a great solution here :-/ Even though the patch isn't used in our MSRV build for the `rustls` lib (it's only used by the `connect-tests` and `examples` crates) we can't resolve the patch unless we're using 1.64+ and so the MSRV task is broken.

Some ideas:

* Wait for an upstream release of hickory-dns with the fixes we need - this seems crummy, we could be waiting for a while.
* Switch our MSRV to 1.64 - this seems crummy (and Debian stable is still on 1.63).
* Switch to using a hickory-dns release before the 1.64+ workspace inheritance change. This would mean backporting my SVCB fixes to trust-dns 0.20 and reverting the connect-test/example code to use the old API.

The last idea seems the least gross, but still not ideal :face_exhaling:

Reply:
> Wait for an upstream release of hickory-dns with the fixes we need - this seems crummy, we could be waiting for a while.

Ctz rightly points out we could make do with a published pre-release. I'll file an issue to inquire about this.

Reply:
> Switch to using a hickory-dns release before the 1.64+ workspace inheritance change. This would mean backporting my SVCB fixes to trust-dns 0.20 and reverting the connect-test/example code to use the old API.

While we're waiting I did a quick-n-dirty version of this (see `cpu/trust-dns` [cpu-0.22-ech-fixes branch](https://github.com/cpu/trust-dns/tree/cpu-0.22.0-ech-fixes) - I backported only the core ECH fixes, not the comment updates or the general SVCB improvements). The rustls side updates to use this compatible `patch` and the older APIs are in a separate commit at the end so that we can hopefully revert it once a pre-release is available.

Doing this uncovered one real MSRV issue with this branch: I was using a `let ... else` construction that's unstable on 1.63. I fixed this in the original commit that introduced it along with a TODO to revert to the nicer construction when we bump MSRV.

With both of the above in place this branch should be all green in CI.